### PR TITLE
Jv vision demo

### PR DESCRIPTION
**Pull request by jvic12**
This project includes two procedures that will demonstrate the VisionScript image processing capabilities in a procedure called vsExample and TensorFlow object recognition support in a procedure called tfExample. The main purpose of this demo is to showcase VANTIQ's integration with TensorFlow and as a result the VisionScript aspect is more of a side project and is a smaller piece of the demo.

A simple client TfVisionDemo has been created that will allow users to choose or upload an image to VANTIQ that will be processed through TensorFlow and the results displayed on the page. The client also allows you to change the confidence threshold and model type for image classification and toggle to the VisionScript part of the demo.

An important thing to note is that this demo does require some additional set up on the users end with loading the TensorFlow models into Vantiq. Further instructions can be found in the project description panel in the IDE once the project has been imported. 

Another important detail about this demo is that errors have popped up that are out of the scope of this demo but can cause it to fail. The primary errors that I have encountered have to do with the model not loading in time, which usually resolves itself just by waiting for the model to finish loading, and one where the resources appear to be unavailable, which has occurred after servers weren't restarted and requires higher level authority to resolve. 

Client Home Page
Vision Script Photo
Example Model JSON

### CODE REVIEW

**Review by JakeSchmitz — CHANGES_REQUESTED** [review]
Before going into the code, I have a few top level thoughts:

It's not clear why there is an extra Pictures type, rather than just querying images directly. I would prefer if the list in the client was populated by querying the images type directly, as the way it works now there are a bunch of images listed that don't actually exist and produce errors.

I'm also concerned about the use of hardcoded image names for blood.jpg in the procedure and the client. I would prefer if this didn't need to be hardcoded anywhere, and could be passed as a parameter to the procedure and could be selected from a list of images in the client. 

The image loading instructions don't seem to work (or there is an issue with the 1.29 CLI), I've followed the instructions listed in the description pane and was able to load a tensorflowmodel using the CLI, but the image loading instructions don't seem to work (and produce no error message either, so clearly something is wrong).

The namespace listed in the CLI command examples is vision-demo, but that isn't a legal namespace name so I got a little confused. I would prefer if the CLI commands used the profile option like `vantiq -s <profileName> load tensorflowmodels coco-1.2` rather than enumerating all the options (like token and namespace name) that will be different for users of this contribution. 

Overall the Contribution looks good, and the description for how to set up tf in vantiq was well written. Most of the above suggestions should be pretty easy to change, but let me know if you want to talk about any of it or if I misunderstood anything.

File: VisionDemo/projects/VisionDemo.json
Comment (rblumer4) [inline VisionDemo/projects/VisionDemo.json]: couple typos in the description here:  
     specifc  -> specific
     yor       -> your
Also, some URLs reference dev.vantiq.com & some use internal.vantiq.com.  They should all be the same, so you should change internal to dev.vantiq.com

**Comment by rblumer4** [comment]
There are no .jpg files in the PR.
Can you add the jpg files referenced in data/Pictures.json, so that they will be imported when the VisionDemo directory is imported?

**Comment by sfitts** [comment]
> There are no .jpg files in the PR.

Depending on their size, it might be better to host these elsewhere and simply reference them.  Large amounts of binary data in github is.... not good.  So if it's a few MB, that's fine, but if it's GBs, then we should look for alternatives.

**Comment by jvic12** [comment]
I decided to remove the VisionScript part of the demo because it made the underlying code a lot more convoluted and it also made the user experience more difficult to get the demo up and running. The TensorFlow part of the demo allows the users to upload images through the client in the form of documents which can be processed by the TensorFlowOperation. However, the VisionScript only can process images, which requires the user to upload images through the Vantiq CLI and also requires both a second data type to be stored and would probably also necessitate sample images being uploaded and downloaded through GitHub, which isn’t ideal. Because of the extra baggage for what isn’t the main focus of the demo, I decided that it is best to hold off on adding this to the Object Recognition Contribution until there is more interchangeability between images and documents. Because of this, the demo will be more lightweight and simpler to use. Here is an updated image of what the client looks like:
Updated Client

File: VisionDemo/projects/TensorFlowDemo.json
Comment (rblumer4) [inline VisionDemo/projects/TensorFlowDemo.json]: there is still one reference to _internal_.vantiq.com that should be changed to _dev.vantiq.com_

Comment (jvic12) [reply VisionDemo/projects/TensorFlowDemo.json]: Thanks for catching that, updating that right now

File: VisionDemo/clients/TFDemo/controllers/dataStream/dc_insert_documents/onDataArrived.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/dataStream/dc_insert_documents/onDataArrived.js]: remove all the console.logs except for logging errors please

File: VisionDemo/clients/TFDemo/controllers/dataStream/dc_insert_documents/onDataArrived.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/dataStream/dc_insert_documents/onDataArrived.js]: indentation needs a little fixing here

File: VisionDemo/clients/TFDemo/controllers/dataStream/dc_insert_documents/onDataArrived.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/dataStream/dc_insert_documents/onDataArrived.js]:
```suggestion
                    ctx.lineWidth ="2";	
                    ctx.strokeStyle = client.getWidget("BoxColor").platformWidget.rawValue;	
```

File: VisionDemo/clients/TFDemo/controllers/page/Start/widgets/RefreshButton/onClick.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/page/Start/widgets/RefreshButton/onClick.js]: Can you switch this to `currDoc`? You've used camelcase in other code for variable naming and that's more consistent with our variable naming than snake case.

File: VisionDemo/clients/TFDemo/controllers/page/Start/widgets/ChangeModelButton/onClick.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/page/Start/widgets/ChangeModelButton/onClick.js]: This click to cycle thing is kinda confusing, and it's going to be a pain to upgrade over time as the most popular models change. Can you change this to be a droplist populated by the tensorflowmodels that you can get by querying on client startup?

File: VisionDemo/clients/TFDemo/controllers/page/Start/widgets/DataTable1/onButtonClickCallback.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/page/Start/widgets/DataTable1/onButtonClickCallback.js]: This code is all duplicated from the onDataArrived event handler for the data stream. I think you should commonize this code into a single function defined on the client.

File: VisionDemo/clients/TFDemo/controllers/page/Start/widgets/RefreshButton/onClick.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/page/Start/widgets/RefreshButton/onClick.js]: ditto to commonizing duplicative code

**Review by JakeSchmitz — CHANGES_REQUESTED** [review]
Overall, this looks very good! I had a lot of fun playing around with it on my catalog of wildlife photos, but unfortunately the coco-1.2 model I was using seems to mostly think everything is a bird. 

I have a few suggestions that are mostly minor, then this should be good to go.

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/client/onStart.js:9]: Just call this function processImage and skip the `processImage = processImageGlobal` at the end of the file

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/client/onStart.js:32]: line width

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/client/onStart.js:37]: line width

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/client/onStart.js:52]: remove

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/client/onStart.js:51]: ;

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [inline VisionDemo/clients/TFDemo/controllers/client/onStart.js:46]: Indentation not consistent here

File: VisionDemo/procedures/tfExample.vail
Comment (JakeSchmitz) [inline VisionDemo/procedures/tfExample.vail:5]: Why not just call TensorFlowOperatipon.processDocument directly from the client and remove this procedure entirely? It's better to direct users to built-ins when possible

**Review by JakeSchmitz — APPROVED** [review]
A few final comments, but otherwise this LGTM! Once those are fixed feel free to merge this in.

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (jvic12) [reply VisionDemo/clients/TFDemo/controllers/client/onStart.js:9]: The line at the bottom is necessary so that the function has a global designation. If I remove it then it throws errors in other files as it is considered a local function.

File: VisionDemo/procedures/tfExample.vail
Comment (jvic12) [reply VisionDemo/procedures/tfExample.vail:5]: I tried calling it from the client and it is undefined so I think that the operation needs to be called from a procedure

File: VisionDemo/clients/TFDemo/controllers/client/onStart.js
Comment (JakeSchmitz) [reply VisionDemo/clients/TFDemo/controllers/client/onStart.js:9]: Does it work instead if you do:

```
var processImage = function(design, args) {
...
}
```

I thought if you used var to declare the function it would be globally visible, but maybe I'm wrong about that.

Comment (jvic12) [reply VisionDemo/clients/TFDemo/controllers/client/onStart.js:9]: I tried changing it to that but still got the same error, I know it's not pretty but I tried some other ways to get a cleaner global function and this is the only one that works.

**Pull request merged**
